ethcore/src/engines/authority_round/mod.rs

```
@@ -1527,6 +1541,15 @@ impl Engine<EthereumMachine> for AuthorityRound {
 			self.validators.report_malicious(header.author(), set_number, header.number(), Default::default());
 			Err(EngineError::DoubleVote(*header.author()))?;
 		}
+		// Report malice if the validator produced other sibling blocks in the same step.
+		let received_block_key = (header.number(), *header.author());
```

Should this actually be the step instead of the number?
If the longest chain changes, it can be valid to create two blocks at the same heights, I think. What we want to report as malicious is to create two blocks in the same _step_, isn't it?

---

Nice catch! You are right, and the comment also implies that we should use the step number.

---

What if a validator produces two blocks with step numbers `n` and `n+m` where `n` is the timestamp of the start of a block and `m` is less than the current step duration? These blocks will have different timestamps and so will be allowed if we change verification to take steps into account instead of block numbers. However the block numbers of those two blocks should be equal.

Do we also check that every block has only one corresponding step?

---

I'm confused: I think a step is different from a timestamp. If two blocks were produced shortly after each other within the period of a single step, then both should have the exact same step number.
> Do we also check that every block has only one corresponding step?

If I understand correctly, the step is just [the first header field](https://github.com/poanetwork/parity-ethereum/blob/52cb3b3fe92bcf207d6588cdf8ff30a709ab7d01/ethcore/src/engines/authority_round/mod.rs#L637), and [the number of header fields is checked already](https://github.com/poanetwork/parity-ethereum/blob/52cb3b3fe92bcf207d6588cdf8ff30a709ab7d01/ethcore/src/engines/authority_round/mod.rs#L628).

---

You are right. Sorry for the confusion! So valid step numbers are an interval of `u64` starting from the first timestamp divided by the first duration and up to `u64::MAX`. There is one-to-one correspondence between step numbers within this interval and block numbers. The latter are a bit easier to think about because a valid block numbers is just any `u64`. What we need to ensure is that we don't accept blocks with step numbers outside the allowed interval.

---

[`verify_timestamp`](https://github.com/poanetwork/parity-ethereum/blob/d00d1ac3c18b12835bd8198dd0512dd33f83ef3f/ethcore/src/engines/authority_round/mod.rs#L691) is perhaps the place to do that.

---

Yes… looks like it's already doing that. Or do you have an additional check in mind?

---

It looks so to me too. I'm adding a unit test for that. Or maybe there already was such a unit test? I didn't find it in any case.

---

...and the new test failed. Not totally sure if the test is implementing my intended check but rather more sure than not sure.

---

`verify_timestamp` is not checking for correspondence between steps and blocks in general. It performs a partial check for the incoming block step numbers to be "close enough" in the future relative to the current step. Blocks with step numbers in the past are OK and don't lead to an error, and that includes any step numbers for which there is no corresponding block. It is possible to successfully receive blocks with such step numbers.

Does this cause a concern? Should we return an error on such step numbers?